### PR TITLE
Game Master flag, preliminary work

### PR DESCRIPTION
**Pull request by pakfront**
This add a new 'gameMaster' (GM) flag to Player. It allows for privileged actions to facilitate refereeing and game mastering of games. At the moment this is limited to allowing /seeall on a GM, even if it is not an observer. More advanced features will be added in the future.
- player using /gm command will start voting for them to be a Game Master
-- other players can vote for this with /allowGM
- player who is already a GM can remove GM mode by using /gm again. No vote is required to disable.
-  tries to rationalize some of the voting code. Probably more can be done, but it does fix the problem where a solo human player cannot trigger a vote to be counted.
- Resolves https://github.com/MegaMek/megamek/issues/3786
- adds more detail to Player List UI (this could use some prettying up later)
- properly copies some flags when Player is cloned
- Resolves https://github.com/MegaMek/megamek/issues/3784
- Consolidates double blind checks against seeAll and singleblind

### CODE REVIEW

**Comment by Thom293** [comment]
This looks great!  I had no idea you were so close.  You are welcome  to manipulate singleblind to add a vote or otherwise work with GM mode, however you wish.

Just a note: in the vanilla game - I could not get /seeall to work with a password, but I just may not know what I am doing.  Or it may not have worked because I was running it through the IDE.  Not sure.  I just could not get it to work.

**Comment by pakfront** [comment]
Thanks. Singleblind functionality looks good as is. I don't think a vote is needed. I will probably clean up a few things here and there and I will test to see if I can get a password to work with seeall

File: megamek/src/megamek/common/Player.java [hunk 1]
```diff
@@ -84,12 +83,10 @@ public final class Player extends TurnOrdered {
...
     private boolean admitsDefeat = false;
-    
-    /**
-     * Boolean that keeps track of whether a player has accepted another 
-     * player's request to change teams.
-     */
-    private boolean allowingTeamChange = false;
+
+    //TODO Voting should not be stored in savegame. mark transient?
+    private boolean votedToAllowTeamChange = false;
```
Comment (NickAragua) [inline megamek/src/megamek/common/Player.java:88]: Yeah, transient is the way to go here.

File: megamek/src/megamek/server/commands/AllowGameMasterCommand.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+package megamek.server.commands;
+
+import megamek.common.Player;
+import megamek.server.GameManager;
+import megamek.server.Server;
+
+/**
+ * This command allows a player to allow another player to switch teams.
+ *
+ * @author arlith
```
Comment (NickAragua) [inline megamek/src/megamek/server/commands/AllowGameMasterCommand.java:28]: Give yourself a little credit. ;)

File: megamek/src/megamek/server/commands/AllowGameMasterCommand.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    protected static void voteYes(Server server, Player player ) {
+        player.setVotedToAllowGameMaster(true);
+
+        // Tally votes
+        boolean allowGameMaster = true;
+        int voteCount = 0;
+        int eligiblePlayerCount = 0;
+        for (Player p : server.getGame().getPlayersVector()) {
+            if (p.getTeam() != Player.TEAM_UNASSIGNED) {
+                allowGameMaster &= p.getVotedToAllowGameMaster();
```
Comment (NickAragua) [inline megamek/src/megamek/server/commands/AllowGameMasterCommand.java:67]: Do we want to shortcut this to break out of the loop whenever we hit a "no" vote?

**Review by NickAragua — CHANGES_REQUESTED** [review]
A little documentation and one minor optimization.

File: megamek/src/megamek/server/commands/AllowGameMasterCommand.java (hunk 3 above)
Comment (pakfront) [reply megamek/src/megamek/server/commands/AllowGameMasterCommand.java:67]: Currently there is no way to vote "no." False just mean you haven't voted yet. So the vote hangs unresolved until someone restarts the vote count with /gm (/joinTeam uses the same logic)

 Adding a "no" option would be a new feature, one worth having. A some generic voting code would be good.

Comment (NickAragua) [reply megamek/src/megamek/server/commands/AllowGameMasterCommand.java:67]: Let's limit scope and leave it for another pull request, then.

**Review by NickAragua — APPROVED** [review]

**Comment by Thom293** [comment]
Nice.   Thank you both.

**Pull request merged**